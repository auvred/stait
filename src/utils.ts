export type ContextDecl = Record<string, unknown>

export type ActionsDecl<_Typestait extends { type: string }> = Record<
  string,
  // `any` looks cleaner in user-facing errors
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (payload?: any) => _Typestait['type']
>

export type Flatten<T> = [T] extends [infer TT]
  ? {
      [K in keyof TT]: TT[K]
    }
  : never
