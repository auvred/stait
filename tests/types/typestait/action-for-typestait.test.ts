import { runTypesTests } from '../../utils'

import type { ActionForTypestait } from '../../../src'
import type { Equal } from '../../utils'

runTypesTests(import.meta.url)

type ExpectType_simple_recursive_action_depth_1 = ActionForTypestait<
  {
    type: 'stait1'
    actions: {
      action1: () => 'stait1'
    }
  },
  'stait1',
  'action1'
>

type ExpectType_simple_recursive_action_depth_2 =
  ReturnType<ExpectType_simple_recursive_action_depth_1>['actions']['action1']

type ExpectType_simple_recursive_action_depth_2_equals_depth_1 = Equal<
  ExpectType_simple_recursive_action_depth_1,
  ExpectType_simple_recursive_action_depth_2
>

type ExpectError_action_for_non_existent_stait = ActionForTypestait<
  {
    type: 'stait1'
    actions: {
      action1: () => 'stait1'
    }
  },
  'stait2',
  'action1'
>

type ExpectError_non_existent_action = ActionForTypestait<
  {
    type: 'stait1'
    actions: {
      action1: () => 'stait1'
    }
  },
  'stait1',
  'action2'
>

type ExpectType_without_context_and_without_payload = ActionForTypestait<
  {
    type: 'stait1'
    actions: {
      action1: () => 'stait1'
    }
  },
  'stait1',
  'action1'
>
type ExpectType_args_without_context_and_without_payload = Equal<
  Parameters<ExpectType_without_context_and_without_payload>,
  []
>

type ExpectType_with_context_and_without_payload = ActionForTypestait<
  {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: () => 'stait1'
    }
  },
  'stait1',
  'action1'
>
type ExpectType_args_with_context_and_without_payload = Equal<
  Parameters<ExpectType_with_context_and_without_payload>,
  [
    {
      ctx: {
        data1: string
        data2: 123
      }
    },
  ]
>

type ExpectType_without_context_and_with_payload = ActionForTypestait<
  {
    type: 'stait1'
    actions: {
      action1: (payload: { data1: string; data2: 123 }) => 'stait1'
    }
  },
  'stait1',
  'action1'
>
type ExpectType_args_without_context_and_with_payload = Equal<
  Parameters<ExpectType_without_context_and_with_payload>,
  [
    {
      payload: {
        data1: string
        data2: 123
      }
    },
  ]
>

type ExpectType_with_context_and_with_payload = ActionForTypestait<
  {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: (payload: { data1: string; data2: 123 }) => 'stait1'
    }
  },
  'stait1',
  'action1'
>
type ExpectType_args_with_context_and_with_payload = Equal<
  Parameters<ExpectType_with_context_and_with_payload>,
  [
    {
      ctx: {
        data1: string
        data2: 123
      }
      payload: {
        data1: string
        data2: 123
      }
    },
  ]
>
