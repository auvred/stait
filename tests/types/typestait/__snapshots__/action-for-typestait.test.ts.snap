// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`error: action for non existent stait 1`] = `
"error TS2344: Type '\\"stait2\\"' does not satisfy the constraint '\\"stait1\\"'.
"
`;

exports[`error: non existent action 1`] = `
"error TS2344: Type '\\"action2\\"' does not satisfy the constraint '\\"action1\\"'.
"
`;

exports[`type: args with context and with payload 1`] = `
"type args_with_context_and_with_payload =
  true
"
`;

exports[`type: args with context and without payload 1`] = `
"type args_with_context_and_without_payload =
  true
"
`;

exports[`type: args without context and with payload 1`] = `
"type args_without_context_and_with_payload =
  true
"
`;

exports[`type: args without context and without payload 1`] = `
"type args_without_context_and_without_payload =
  true
"
`;

exports[`type: simple recursive action depth 1 1`] = `
"type simple_recursive_action_depth_1 =
  () => {
    type: 'stait1'
    actions: {
      action1: any
    }
  }
"
`;

exports[`type: simple recursive action depth 2 1`] = `
"type simple_recursive_action_depth_2 =
  () => {
    type: 'stait1'
    actions: {
      action1: any
    }
  }
"
`;

exports[`type: simple recursive action depth 2 equals depth 1 1`] = `
"type simple_recursive_action_depth_2_equals_depth_1 =
  true
"
`;

exports[`type: with context and with payload 1`] = `
"type with_context_and_with_payload =
  (opts: {
    ctx: {
      data1: string
      data2: 123
    }
    payload: {
      data1: string
      data2: 123
    }
  }) => {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: any
    }
  }
"
`;

exports[`type: with context and without payload 1`] = `
"type with_context_and_without_payload =
  (opts: {
    ctx: {
      data1: string
      data2: 123
    }
  }) => {
    type: 'stait1'
    context: {
      data1: string
      data2: 123
    }
    actions: {
      action1: any
    }
  }
"
`;

exports[`type: without context and with payload 1`] = `
"type without_context_and_with_payload =
  (opts: {
    payload: {
      data1: string
      data2: 123
    }
  }) => {
    type: 'stait1'
    actions: {
      action1: any
    }
  }
"
`;

exports[`type: without context and without payload 1`] = `
"type without_context_and_without_payload =
  () => {
    type: 'stait1'
    actions: {
      action1: any
    }
  }
"
`;
